logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/LocalRequest.java

```diff
@@ -13,20 +16,126 @@
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static com.squareup.okhttp.HttpUrl.defaultPort;
 import static com.squareup.okhttp.RequestBody.create;
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.zalando.fauxpas.FauxPas.throwingUnaryOperator;
 
+@AllArgsConstructor
 final class LocalRequest implements HttpRequest {
 
-    private Request request;
-    private byte[] body;
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer() throws IOException {
+            return this;
+        }
+
+        Request getRequest();
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    @AllArgsConstructor
+    private abstract class AbstractState implements State {
+
+        @Override
+        public Request getRequest() {
+            return request;
+        }
+
+    }
+
+    private final class Unbuffered extends AbstractState {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private final class Offering extends AbstractState {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer() throws IOException {
+            @Nullable final RequestBody entity = request.body();
+
+            if (entity == null) {
+                return new Passing();
+            } else {
+                final Buffer buffer = new Buffer();
+                entity.writeTo(buffer);
+                final byte[] body = buffer.readByteArray();
+
+                final Request copy = request.newBuilder()
+                        .method(request.method(), create(entity.contentType(), body))
+                        .build();
+
+                return new Buffering(copy, body);
+            }
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Buffering implements State {
+
+        @Getter
+        private final Request request;
+        private final byte[] body;
+
+        @Override
+        public State without() {
+            return new Ignoring(request, body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Ignoring implements State {
+
+        @Getter
+        private final Request request;
+        private final byte[] body;
+
+        @Override
+        public State with() {
+            return new Buffering(request, body);
+        }
+
+    }
+
+    private final class Passing extends AbstractState {
 
-    LocalRequest(final Request request) {
-        this.request = request;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)